src/middleware.ts

```diff
@@ -259,43 +296,60 @@ export const persist = <S extends State>(
   }
 
   // rehydrate initial state with existing stored state
-  ;(async () => {
-    const postRehydrationCallback = onRehydrateStorage?.(get()) || undefined
 
-    try {
-      const storageValue = await storage.getItem(name)
-
-      if (storageValue) {
-        const deserializedStorageValue = await deserialize(storageValue)
-
-        // if versions mismatch, run migration
-        if (deserializedStorageValue.version !== version) {
-          const migratedState = await migrate?.(
-            deserializedStorageValue.state,
-            deserializedStorageValue.version
-          )
-          if (migratedState) {
-            set(migratedState)
-            await setItem()
+  // The state that was loaded from the storage and is set via set() within this
+  // middleware would later be overridden with the initial state by create(). To
+  // avoid this we merge the state from localStorage into the initial state. This
+  // way the state gets overridden with the state loaded from storage.
+  // TODO: find a better solution for this
+  let stateFromStorage
+  ;(() => {
+    const postRehydrationCallback = onRehydrateStorage?.(get()) || undefined
+    // bind is used to avoid `TypeError: Illegal invocation` error
+    let storageValuePending = toThenable(storage.getItem.bind(storage))(name)
+    storageValuePending
+      .then((storageValue) => {
+        if (storageValue) {
+          return deserialize(storageValue)
+        }
+      })
+      .then((deserializedStorageValue) => {
+        if (deserializedStorageValue) {
+          if (deserializedStorageValue.version !== version) {
+            if (!migrate) return
+            stateFromStorage = migrate(
+              deserializedStorageValue.state,
+              deserializedStorageValue.version
+            )
+            return stateFromStorage
+          } else {
+            stateFromStorage = deserializedStorageValue.state
+            set(deserializedStorageValue.state)
           }
-        } else {
-          set(deserializedStorageValue.state)
         }
-      }
-    } catch (e) {
-      postRehydrationCallback?.(undefined, e)
-      return
-    }
-
-    postRehydrationCallback?.(get(), undefined)
+      })
+      .then((migratedState) => {
+        if (migratedState) {
+          set(migratedState as PartialState<S, keyof S>)
+          return setItem()
+        }
+      })
+      .then(() => {
+        postRehydrationCallback?.(get(), undefined)
+      })
+      .catch((e: Error) => {
+        postRehydrationCallback?.(undefined, e)
+      })
   })()
 
-  return config(
+  const configResult = config(
     (...args) => {
       set(...args)
       void setItem()
     },
     get,
     api
   )
+
+  return { ...configResult, ...(stateFromStorage || {}) }
```

Comment:
This is the workaround, right? I'm not sure if I fully understand it.

```suggestion
  return stateFromStorage ? configResult : { ...configResult, ...stateFromStorage }
```
should be better.

Reply:
Would need to be the other way around:
```suggestion
  return stateFromStorage ? { ...configResult, ...stateFromStorage } : configResult
```

It's the continuation of what's described in the commend above:
```js
  // The state that was loaded from the storage and is set via set() within this
  // middleware would later be overridden with the initial state by create(). To
  // avoid this we merge the state from localStorage into the initial state. This
  // way the state gets overridden with the state loaded from storage.
```

Reply:
yeah, it's a workaround.
I think there is something wrong with provided snipped.